### PR TITLE
Resize

### PR DESCRIPTION
**Pull request by ibeckermayer**
This adds resize capabilities to Teleport. This required several changes to IronRDP:

- https://github.com/Devolutions/IronRDP/pull/418
- https://github.com/Devolutions/IronRDP/pull/419
- https://github.com/Devolutions/IronRDP/pull/424
- https://github.com/Devolutions/IronRDP/pull/430
- https://github.com/Devolutions/IronRDP/pull/436

TODO:
- https://github.com/gravitational/teleport/pull/39895
- [x] Change the ironrdp rev hash in Cargo.toml to reflect `master` once the PRs above are merged.
- [x] Create a separate PR with any changes required to get session recordings working. Update: no changes were needed.
- [x] Create a separate PR with changes required to update Rust and implement the [suggested workaround](https://github.com/rust-lang/rust/pull/122847) to get Arm builds to succeed. Update: https://github.com/gravitational/teleport/pull/39895
- [x] Investigate the known issue of performance beginning to degrade after several resizes.

changelog: Adds resize capability to windows desktop sessions.

### CODE REVIEW

File: .github/workflows/os-compatibility-test.yaml
Comment (zmb3) [inline .github/workflows/os-compatibility-test.yaml:1]: Why are these changes necessary? Doesn't seem related to resizing.

File: e
Comment (zmb3) [inline e:1]: Is the e update intentional here?

File: lib/srv/desktop/rdp/rdpclient/Cargo.toml [hunk 1]
```diff
@@ -36,6 +38,8 @@ tokio = { version = "1.37", features = ["full"] }
 tokio-boring = { git = "https://github.com/gravitational/boring", rev="605253d99d5e363e178bcf97e1d4622e33844cd5", optional = true }
 utf16string = "0.2.0"
 uuid = { version = "1.8.0", features = ["v4"] }
+tracing-subscriber = "0.3.17"
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/Cargo.toml:41]: Are these new crates necessary?

File: lib/srv/desktop/rdp/rdpclient/src/client.rs [hunk 2]
```diff
@@ -65,6 +74,10 @@ use tokio_boring::{HandshakeError, SslStream};
 
 const RDP_CONNECT_TIMEOUT: Duration = Duration::from_secs(5);
 
+struct ResizeWithholder {
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/client.rs:77]: Maybe a comment on the purpose of this utility would help (ie why we need to withhold resizes)

File: lib/srv/desktop/rdp/rdpclient/src/client.rs [hunk 3]
```diff
@@ -149,8 +164,19 @@ impl Client {
...
+        let display_control = DisplayControlClient::new(move |_| {
+            Self::on_display_ctl_capabilities_received(&resize_manager_clone)
+        });
+        let drdynvc_client = DrdynvcClient::new().with_dynamic_channel(display_control);
+
+        let mut connector = ironrdp_connector::ClientConnector::new(connector_config.clone())
             .with_server_addr(server_socket_addr)
+            .with_static_channel(drdynvc_client)
             .with_static_channel(Rdpsnd::new()) // required for rdpdr to work
             .with_static_channel(rdpdr);
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/client.rs:181]:
```suggestion
            .with_static_channel(drdynvc_client) // require for resizing
            .with_static_channel(Rdpsnd::new()) // required for rdpdr to work
            .with_static_channel(rdpdr); // required for smart card + directory sharing
```

File: web/packages/teleport/src/DesktopSession/useTdpClientCanvas.tsx [hunk 4]
```diff
@@ -282,6 +272,11 @@ export default function useTdpClientCanvas(props: Props) {
   // on the remote machine.
   const canvasOnContextMenu = () => false;
 
+  const windowOnResize = throttle((cli: TdpClient) => {
```
Comment (zmb3) [inline web/packages/teleport/src/DesktopSession/useTdpClientCanvas.tsx:275]: I think `debounce` would be a better choice than `throttle` here.

https://web.archive.org/web/20220714020647/https://bencentra.com/code/2015/02/27/optimizing-window-resize.html

> debouncing window.onresize will only call the event handler after the event has stopped firing for a certain amount of time. This will ensure that your function will only be called once the resizing is “complete.”

File: web/packages/teleport/src/lib/tdp/client.ts [hunk 5]
```diff
@@ -325,12 +325,12 @@ export default class Client extends EventEmitterWebAuthnSender {
     );
   }
 
-  handleRDPConnectionInitialized(buffer: ArrayBuffer) {
+  handleRDPConnectionActivated(buffer: ArrayBuffer) {
```
Comment (zmb3) [inline web/packages/teleport/src/lib/tdp/client.ts:328]:
```suggestion
  handleRdpConnectionActivated(buffer: ArrayBuffer) {
```

Here and below (we don't capitalize initialisms in JS code the way we would in Go)

File: lib/srv/desktop/rdp/rdpclient/src/client.rs [hunk 6]
```diff
@@ -74,6 +87,7 @@ pub struct Client {
     write_stream: Option<RdpWriteStream>,
     function_receiver: Option<FunctionReceiver>,
     x224_processor: Arc<Mutex<x224::Processor>>,
+    resize_withholder: Arc<Mutex<ResizeWithholder>>,
```
Comment (probakowski) [inline lib/srv/desktop/rdp/rdpclient/src/client.rs:90]: Maybe call it `pending_resize`? 
And maybe we can go with plain `Arc<Mutex<Option<(u32,u32)>>>`, or 
`type PendingResize = Option<(u32,u32)>; .... Arc<Mutex<PendingResize>>`

File: lib/srv/desktop/rdp/rdpclient/src/client.rs [hunk 7]
```diff
@@ -149,8 +164,19 @@ impl Client {
             debug!("creating rdpdr client with directory sharing disabled")
         }
 
-        let mut connector = ironrdp_connector::ClientConnector::new(connector_config)
+        let resize_withholder = Arc::new(Mutex::new(ResizeWithholder {
+            withheld_resize: None,
+        }));
+
+        let resize_manager_clone = resize_withholder.clone();
```
Comment (probakowski) [inline lib/srv/desktop/rdp/rdpclient/src/client.rs:171]: Looks like we have some discrepancy with naming here

File: lib/srv/desktop/rdp/rdpclient/Cargo.toml (hunk 1 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/Cargo.toml:41]: Hmm, I don't think so. @probakowski it looks like you may have added these before I took over for you, do we need to keep them for some reason?

File: .github/workflows/os-compatibility-test.yaml
Comment (ibeckermayer) [reply .github/workflows/os-compatibility-test.yaml:1]: That may have been a mistake from a merge, reverted.

File: e
Comment (ibeckermayer) [reply e:1]: That may have been a mistake from a merge, reverted.

File: web/packages/teleport/src/lib/tdp/client.ts [hunk 8]
```diff
@@ -325,12 +325,12 @@ export default class Client extends EventEmitterWebAuthnSender {
     );
   }
 
-  handleRDPConnectionInitialized(buffer: ArrayBuffer) {
+  handleRdpConnectionActivated(buffer: ArrayBuffer) {
     const { ioChannelId, userChannelId, screenWidth, screenHeight } =
-      this.codec.decodeRDPConnectionInitialied(buffer);
+      this.codec.decodeRDPConnectionActivated(buffer);
```
Comment (zmb3) [inline web/packages/teleport/src/lib/tdp/client.ts:330]:
```suggestion
      this.codec.decodeRdpConnectionActivated(buffer);
```

File: web/packages/teleport/src/lib/tdp/client.ts [hunk 9]
```diff
@@ -343,7 +343,7 @@ export default class Client extends EventEmitterWebAuthnSender {
     this.emit(TdpClientEvent.TDP_CLIENT_SCREEN_SPEC, spec);
   }
 
-  handleRDPFastPathPDU(buffer: ArrayBuffer) {
+  handleRdpFastPathPDU(buffer: ArrayBuffer) {
```
Comment (zmb3) [inline web/packages/teleport/src/lib/tdp/client.ts:346]:
```suggestion
  handleRdpFastPathPdu(buffer: ArrayBuffer) {
```

File: web/packages/teleport/src/lib/tdp/codec.ts [hunk 10]
```diff
@@ -108,7 +108,7 @@ export type ClipboardData = {
 };
 
 // | message type (31) | io_channel_id uint16 | user_channel_id uint16 | screen_width uint16 | screen_height uint16 |
-export type RDPConnectionInitialized = {
+export type RDPConnectionActivated = {
```
Comment (zmb3) [inline web/packages/teleport/src/lib/tdp/codec.ts:111]:
```suggestion
export type RdpConnectionActivated = {
```

File: web/packages/teleport/src/lib/tdp/codec.ts [hunk 11]
```diff
@@ -875,7 +875,7 @@ export default class Codec {
   }
 
   // | message type (31) | io_channel_id uint16 | user_channel_id uint16 | screen_width uint16 | screen_height uint16 |
-  decodeRDPConnectionInitialied(buffer: ArrayBuffer): RDPConnectionInitialized {
+  decodeRDPConnectionActivated(buffer: ArrayBuffer): RDPConnectionActivated {
```
Comment (zmb3) [inline web/packages/teleport/src/lib/tdp/codec.ts:878]:
```suggestion
  decodeRdpConnectionActivated(buffer: ArrayBuffer): RdpConnectionActivated {
```

**Review by zmb3 — APPROVED** [review]
Looks good aside from inconsistent capitalization of acronyms.

**Review by probakowski — APPROVED** [review]
